ui-v2/app/components/data-source.js

```diff
@@ -0,0 +1,121 @@
+ * not the component.
+ *
+ * @example ```js
+ *   {{data-source
+ *      src="/dc-1/services/*"
+ *      onchange={{action (mut items) value='data'}}
+ *      onerror={{action (mut error) value='error'}}
+ *   /}}```
+ *
+ * @param src {String} - An identitier used to determine the source of the data. This is passed
+ *                       to the data service for it to determine how to fetch the data.
+ * @param onchange=null {Func} - An action called when the data changes.
+ * @param onerror=null {Func} - An action called on error
+ *
+ */
+export default Component.extend(WithListeners, {
+  tagName: 'span',
+
+  // TODO: can be injected with a simpler non-blocking
+  // data service if we turn off blocking queries completely at runtime
+  data: service('blocking'),
+
+  onchange: function() {},
+  onerror: function() {},
+
+  didInsertElement: function() {
+    this._super(...arguments);
+    const options = {
+      rootMargin: '0px',
+      threshold: 1.0,
+    };
+
+    const observer = new IntersectionObserver((entries, observer) => {
+      entries.map(item => {
+        set(this, 'isIntersecting', item.isIntersecting);
+        if (!item.isIntersecting) {
+          this.actions.close.bind(this)();
+        } else {
+          this.actions.open.bind(this)();
+        }
+      });
+    }, options);
+    observer.observe(this.element); // eslint-disable-line ember/no-observers
+    this.listen(() => {
+      this.actions.close.bind(this)();
+      observer.disconnect(); // eslint-disable-line ember/no-observers
+    });
+  },
+  didReceiveAttrs: function() {
+    this._super(...arguments);
+    if (this.element && this.isIntersecting) {
+      this.actions.open.bind(this)();
+    }
+  },
+  actions: {
+    // keep this argumentless
+    open: function() {
+      const src = this.src;
+      const filter = this.filter;
+
+      // get a new source and replace the old one, cleaning up as we go
+      const source = replace(
+        this,
+        'source',
+        this.data.open(`${src}${filter ? `?filter=${filter}` : ``}`, this),
+        (prev, source) => {
+          // Makes sure any previous source (if different) is ALWAYS closed
+          this.data.close(prev, this);
+        }
+      );
+      // set up the listeners (which auto cleanup on component destruction)
+      const remove = this.listen(source, {
+        message: e => this.onchange(e),
+        error: e => this.onerror(e),
+      });
+      replace(this, '_remove', remove);
+      // dispatch the current data of the source if we have any
+      if (typeof source.getCurrentEvent === 'function') {
+        const currentEvent = source.getCurrentEvent();
+        if (currentEvent) {
+          this.onchange(currentEvent);
+        }
+      }
+    },
+    // keep this argumentless
+    close: function() {
+      this.data.close(this.source, this);
+      replace(this, '_remove', null);
+      set(this, 'source', undefined);
```

Comment:
Hmm 🤔 good point, I'm not sure. Will have a look at this just incase, but I think I'll make them both null, I generally don't like settings things to `undefined` as it just sounds wrong (define something as undefined!?), so I'm wondering why I did it here - it could just be a remnant of a previous iteration/prototype thats stuck around - will have a proper look later 👍